### ISSUE TITLE
LogProbAlgebra.join gives a NaN gradient for an all -inf row, where TropicalAlgebra.join does not

### ISSUE THREAD

**Issue by aaronstevenwhite** [FACTSlab/quivers#58]
### Summary

`LogProbAlgebra.join` returns the correct `-inf` for a row whose entries are all `-inf`, but its backward pass fills that row's gradient with `NaN`. In a differentiable dynamic program an impossible cell is exactly an all-`-inf` row, so this turns a legitimate state into a `NaN` that then spreads through the whole gradient.

`TropicalAlgebra.join` handles the same input correctly, so the two algebras disagree on a case both meet.

### Reproduction

```python
import torch
from quivers import LOG_PROB

x = torch.full((2, 3), float("-inf"), requires_grad=True)
out = LOG_PROB.join(x, dim=1)
out.sum().backward()

print(out)      # tensor([-inf, -inf])   correct
print(x.grad)   # tensor([[nan, nan, nan], [nan, nan, nan]])
```

A mixed batch shows the shape of the problem: the finite row is fine and only the impossible row is poisoned.

```python
z = torch.tensor([[0.0, -1.0, float("-inf")],
                  [float("-inf")] * 3], requires_grad=True)
LOG_PROB.join(z, dim=1).sum().backward()

z.grad[0]   # tensor([0.7311, 0.2689, 0.0000])   correct
z.grad[1]   # tensor([nan, nan, nan])
```

For contrast, on the same input:

```python
from quivers import TROPICAL
t = torch.full((2, 3), float("-inf"), requires_grad=True)
TROPICAL.join(t, dim=1).sum().backward()
torch.isnan(t.grad).any()   # tensor(False)
```

quivers 0.17, torch 2.14.0, Python 3.14.

### Cause

`LogProbAlgebra.join` delegates straight to `torch.logsumexp`:

```python
def join(self, t: torch.Tensor, dim: int | tuple[int, ...]) -> torch.Tensor:
    if isinstance(dim, int):
        dim = (dim,)
    return torch.logsumexp(t, dim=dim)
```

`torch.logsumexp` has the same behaviour on its own, so this is inherited rather than introduced. The subtraction of the row maximum leaves `-inf - -inf`, and the derivative of `log` at zero is unbounded, so the row differentiates to `NaN` even though its value is a finite fact about the problem.

### Suggested fix

Guard the row maximum and, more importantly, replace the zero **before** the `log` rather than after it, so `log(0)` never enters the graph:

```python
def join(self, t: torch.Tensor, dim: int | tuple[int, ...]) -> torch.Tensor:
    if isinstance(dim, int):
        dim = (dim,)
    peak = t.amax(dim=dim, keepdim=True)
    shift = peak.masked_fill(peak == float("-inf"), 0.0)
    total = (t - shift).exp().sum(dim=dim)
    empty = total == 0.0
    combined = total.masked_fill(empty, 1.0).log() + shift.squeeze(dim)
    return combined.masked_fill(empty, float("-inf"))
```

Verified on both cases above: the forward values are unchanged (`-inf` for the impossible row, `0.3133` for the mixed one) and neither gradient contains a `NaN`.

Masking after the `log` is not enough. `masked_fill` on the output still differentiates through `log(0)` upstream, and the `NaN` survives.

### Why it matters here

I hit this in a semiring-parameterised inside algorithm, where `LOG_PROB` and `TROPICAL` are the natural log-inside and Viterbi algebras. An all-`-inf` row is the ordinary encoding of a span no rule derives, so the case arises on real inputs rather than at an edge. I ended up writing the guarded reduction locally, which duplicates what `LOG_PROB` is for; with this fixed the local copy can go away.

One note on testing it: a row that is fully **masked** does not expose the bug, because `masked_fill`'s backward zeroes the `NaN`. It takes a row whose unmasked entries are all `-inf`, which is why it can sit unnoticed behind a masked reduction.
